samcli/commands/build/build_context.py

```diff
@@ -307,6 +311,99 @@ def run(self):
             wrapped_from = deep_wrap if deep_wrap else ex.__class__.__name__
             raise UserException(str(ex), wrapped_from=wrapped_from) from ex
 
+    def _enable_source_maps(self):
+        """
+        Appends the --enable-source-maps NODE_OPTIONS if Sourcemap is set to true and vice versa
+        """
+        using_source_maps = False
+        incorrect_node_option = False
+
+        for stack in self.stacks:
+            for _, resource in stack.resources.items():
+                metadata = resource.get("Metadata", {})
+                if metadata.get("BuildMethod", "") != "esbuild":
+                    continue
+
+                node_option_set = self._is_node_option_set(resource)
+
+                # check if Sourcemap is provided and append --enable-source-map if not set
+                build_properties = metadata.get("BuildProperties", {})
+                source_map = build_properties.get("Sourcemap", None)
+
+                if source_map and not node_option_set:
+                    LOG.debug("Sourcemap set without --enable-source-maps, adding")
+
+                    resource.setdefault("Properties", {})
+                    resource["Properties"].setdefault("Environment", {})
+                    resource["Properties"]["Environment"].setdefault("Variables", {})
+                    existing_options = resource["Properties"]["Environment"]["Variables"].setdefault("NODE_OPTIONS", "")
+
+                    # make sure the NODE_OPTIONS is a string
+                    if not isinstance(existing_options, str):
+                        incorrect_node_option = True
+                    else:
+                        resource["Properties"]["Environment"]["Variables"]["NODE_OPTIONS"] = " ".join(
+                            [existing_options, "--enable-source-maps"]
+                        )
+
+                    using_source_maps = True
+
+                # check if --enable-source-map is provided and append Sourcemap: true if it is not set
+                if source_map is None and node_option_set:
+                    LOG.debug("--enable-source-maps set without Sourcemap, adding")
+
+                    resource.setdefault("Metadata", {})
+                    resource["Metadata"].setdefault("BuildProperties", {})
+                    resource["Metadata"]["BuildProperties"]["Sourcemap"] = True
+
+                    using_source_maps = True
+
+        if using_source_maps:
+            self._warn_using_source_maps()
+
+        if incorrect_node_option:
```

Comment:
nit: change incorrect to invalid?, because its just missing a flag right?

Reply:
Its actually when the user provides an invalid NODE_OPTIONS, the customer is able to provide an array of strings in the template file when we are expecting a space delimited string. Still, "invalid" could be used in place of "incorrect" too.